File: langs/typescript/src/services/nullMessageHub.ts
/**
 * NullMessageHub — null-object variant of IMessageHub.
 *
 * See spec/03-messages.md §"Null variant" and ADR-0017.
 */
import { EMPTY, type Observable } from "rxjs";
import type { IMessage } from "../messages/types.js";
import type { ITransactionalMessageHub } from "./messageHub.js";

export class NullMessageHub implements ITransactionalMessageHub {
  /** Shared singleton instance (the hub holds no state). */
  static readonly INSTANCE: NullMessageHub = new NullMessageHub();

  private constructor() {}

  /** Empty observable — completes immediately upon subscribe. */
  readonly messages: Observable<IMessage> = EMPTY;

  /** No-op. */
  send(_message: IMessage): void {
    // intentional no-op per ADR-0017
  }

  /** Execute a transaction body while continuing to publish nothing. */
  batch(transaction: () => void): void {
    transaction();
  }
}

File: langs/swift/Sources/VMx/Services/NullMessageHub.swift
//
// NullMessageHub — null-object variant of `MessageHubProtocol`.
//
// See spec/03-messages.md §"Null variant" and ADR-0017.
//
import Foundation
import Combine

public final class NullMessageHub: TransactionalMessageHubProtocol, @unchecked Sendable {
    /// Shared singleton instance. The hub is stateless.
    public static let INSTANCE = NullMessageHub()

    private init() {}

    /// Empty publisher — completes immediately upon subscribe.
    public let messages: AnyPublisher<any Message, Never> =
        Empty<any Message, Never>(completeImmediately: true)
            .eraseToAnyPublisher()

    /// No-op send per ADR-0017.
    public func send(_ message: any Message) {
        // intentional no-op
    }

    /// Execute the transaction body while continuing to publish nothing.
    public func batch(_ transaction: () throws -> Void) throws {
        try transaction()
    }
}
